### PR TITLE
Feat/contract resizing

### PR DESCRIPTION
**Pull request by leeftk**
# Funding Pot Contract Major Refactor

## Summary
This PR represents a major refactoring of the Funding Pot contract, focusing on simplification, gas optimization, and removal of unnecessary complexity.

### 1. Function Removals
- Removed `getUserEligibility` function (288-344)
  - This was redundant as eligibility is checked during contribution
  - Removed associated interface definition
  - Simplified testing by removing direct eligibility checks

- Removed overloaded `contributeToRoundFor` function (582-596)
  - Consolidated into single function with unspent caps parameter
  - All calls now require unspent caps array (empty if not using)

### 2. Error Handling Consolidation
- Combined multiple round validation errors into `RoundParamsInvalid`:
  - `RoundStartMustBeInFuture`
  - `RoundMustHaveEndTimeOrCap`
  - `RoundEndMustBeAfterStart`
  - `RoundAlreadyStarted`
- Removed parameters from `StartRoundGreaterThanRoundCount` error
- Removed `InvalidBatchParameters` error
- Removed `UnspentCapsRoundIdsNotStrictlyIncreasing`
- Removed `UnspentCapsRoundIdsNotContiguous`

### 3. Accumulation Logic Simplification
- Removed contiguous round ID requirement (659-739)
  - Deleted checks for strictly increasing round IDs
  - Removed validation for sequential round numbers
- Simplified unspent cap tracking:
  - Moved cap marking to main accumulation loop
  - Removed separate marking block after accumulation
- Added early continue conditions for invalid rounds
- Removed redundant round ID validation

### 4. Event and Parameter Optimization
- Simplified `AllowlistedAddressesRemoved` event
  - Removed all parameters (roundId_, accessCriteriaId_, addressesRemoved_)
  - Updated all event emissions
- Changed `merkleProof_` parameter from `calldata` to `memory` in interface
- Removed redundant local variables:
  - Removed `currentTime` in favor of direct `block.timestamp`
  - Removed `r1UnusedTotal` calculation
  - Removed `r1PersonalCap` duplicate

### 5. Security Improvements
- Removed zero address check in `_checkNftOwnership` (1050-1053)
  - Check was redundant due to upstream validations
  - Improves gas efficiency
- Simplified batch processing validation
  - Removed explicit zero check
  - Maintained core functionality with fewer checks

### 6. Test Suite Updates
- Added `_unspentPersonalRoundCaps` as class variable
- Updated all contribution test calls with empty unspent caps array
- Commented out obsolete tests:
  - `testContributeToRoundFor_revertsGivenUnspentCapsWithNonContiguousRoundIds`
  - `testContributeToRoundFor_revertsGivenUnspentCapsRoundIdsNotStrictlyIncreasing`
  - `testContributeToRoundFor_allModeWithGlobalStartRestrictsTotalAccumulation`
- Updated error assertions to use new consolidated errors
- Removed redundant eligibility checks in tests

### 7. Gas Optimizations
- Removed duplicate storage reads
- Consolidated multiple checks into single operations
- Simplified loop conditions
- Removed redundant state tracking
- Optimized event parameter storage

### 8. Interface Changes
```solidity
// Old
function contributeToRoundFor(
    address user_,
    uint32 roundId_,
    uint amount_,
    uint8 accessCriteriaId_,
    bytes32[] calldata merkleProof_
) external;

// New (Single consolidated function)
function contributeToRoundFor(
    address user_,
    uint32 roundId_,
    uint amount_,
    uint8 accessCriteriaId_,
    bytes32[] memory merkleProof_,
    UnspentPersonalRoundCap[] calldata unspentPersonalRoundCaps_
) external;
```

### For Contract Users
1. Update all contribution calls to include empty unspent caps array:
```solidity
// Old
fundingPot.contributeToRoundFor(user, roundId, amount, accessId, proof);

// New
fundingPot.contributeToRoundFor(
    user, 
    roundId, 
    amount, 
    accessId, 
    proof,
    new UnspentPersonalRoundCap[](0)
);
```

2. Update error handling:
```solidity
// Old
try {
    // ... contribution code ...
} catch (bytes memory lowLevelData) {
    if (abi.decodeWithSelector(lowLevelData) == Module__LM_PC_FundingPot__RoundStartMustBeInFuture.selector) {
        // handle specific error
    }
}

// New
try {
    // ... contribution code ...
} catch (bytes memory lowLevelData) {
    if (abi.decodeWithSelector(lowLevelData) == Module__LM_PC_FundingPot__RoundParamsInvalid.selector) {
        // handle consolidated error
    }
}
```

3. Remove any dependency on `getUserEligibility`:
```solidity
// Old
(bool eligible, uint remaining) = fundingPot.getUserEligibility(roundId, accessId, proof, user);

// New
// Implement your own eligibility check or handle it during contribution
```

### For Contract Implementers
1. Update error handling to use consolidated errors
2. Remove any assumptions about round ID ordering in unspent caps
3. Update event handling for `AllowlistedAddressesRemoved`
4. Update tests to use new function signatures and error types

## Testing
- All existing tests have been updated and pass

## Security Considerations
- Removed redundant checks should not impact security
- More tests should be written now that these checks are removed to make sure unspent caps can't be reused
- Tests and functions need to be validated for bugs and edge cases

### CODE REVIEW

File: src/modules/logicModule/LM_PC_FundingPot_v1.sol [hunk 1]
```diff
@@ -807,38 +650,100 @@ contract LM_PC_FundingPot_v1 is
...
+                currentProcessingRoundId,
+                roundCapInfo.accessCriteriaId,
+                roundCapInfo.merkleProof,
+                user_
+            )) {
+                AccessCriteriaPrivileges storage privileges = roundIdToAccessCriteriaIdToPrivileges[currentProcessingRoundId][roundCapInfo.accessCriteriaId];
+                totalAggregatedPersonalCap += privileges.personalCap;
+            }
+            // Mark the specific caps that were used in this contribution
+            usedUnspentCaps[user_][currentProcessingRoundId][roundCapInfo.accessCriteriaId] = true;
```
Comment (omega-audits) [inline src/modules/logicModule/LM_PC_FundingPot_v1.sol:710]: So this will "spend" the full allowance left even if there was more allowance to use. This might be acceptable if the user is aware and in control, but if anyone can contribute on behalf of a user, this allows them to contribute the minimum but still "spend" the full allowance the user had.
